### PR TITLE
Enable endpoints in LXD provider.

### PR DESCRIPTION
**Pull request by Pekkari**
These changes propose to unlock the endpoint configuration in the lxd provider to be able to use an endpoint that is not a gateway. It is an initial work needed to address https://bugs.launchpad.net/juju/+bug/1618798.

### CODE REVIEW

**Comment by Pekkari** [comment]
!!build!!

**Comment by Pekkari** [comment]
!!retry!!

**Comment by Pekkari** [comment]
!!retry!!

File: provider/lxd/environ_raw.go [hunk 1]
```diff
@@ -112,7 +107,7 @@ func newClient(
 // getRemoteConfig returns a lxdclient.Config using a TCP-based remote
 // if called from within an instance started by the LXD provider. Otherwise,
 // it returns an errors satisfying errors.IsNotFound.
-func getRemoteConfig(readFile readFileFunc, runCommand runCommandFunc) (*lxdclient.Config, error) {
+func getRemoteConfig(readFile readFileFunc, runCommand runCommandFunc, spec environs.CloudSpec) (*lxdclient.Config, error) {
```
Comment (axw) [inline provider/lxd/environ_raw.go]: This breaks some tests, which are calling getRemoteConfig directly. Can you please fix them?

Comment (Pekkari) [inline provider/lxd/environ_raw.go]: I'll get this ready, thanks for pointing!

**Comment by Pekkari** [comment]
!!retry!!

**Comment by Pekkari** [comment]
!!retry!!

**Comment by Pekkari** [comment]
!!build!!

**Comment by bz2** [comment]
!!build!!

**Comment by axw** [comment]
@Pekkari Not sure why the bot isn't testing this, but there's a failure in ProviderFunctionalSuite.TestPrepareConfigNonEmptyEndpoint. Would you please remove that one, and add the following test to environ_raw_test.go:

```go
func (s *environRawSuite) TestGetRemoteConfigWithEndpoint(c *gc.C) {
        spec := s.spec
        spec.Endpoint = "1.2.3.4"
        cfg, err := getRemoteConfig(s.readFile, s.runCommand, spec)
        c.Assert(err, jc.ErrorIsNil)
        c.Assert(cfg, jc.DeepEquals, &lxdclient.Config{
                Remote: lxdclient.Remote{
                        Name:     "remote",
                        Host:     "1.2.3.4",
                        Protocol: "lxd",
                        Cert: &lxdclient.Cert{
                                CertPEM: []byte("content:/etc/juju/lxd-client.crt"),
                                KeyPEM:  []byte("content:/etc/juju/lxd-client.key"),
                        },
                        ServerPEMCert: "content:/etc/juju/lxd-server.crt",
                },
        })
}
```

**Comment by Pekkari** [comment]
Ready. Thanks!

**Comment by axw** [comment]
Thanks again.

$$merge$$

**Comment by bz2** [comment]
$$merge$$

**Comment by Pekkari** [comment]
!!retry!!

**Comment by bz2** [comment]
$$merge$$

**Comment by bz2** [comment]
$$retry$$

```
FAIL: uniter_test.go:673: UniterSuite.TestUniterSteadyStateUpgradeRelations
...
util_test.go:136:
    c.Assert(err, jc.ErrorIsNil)
... value *errors.Err = &errors.Err{message:"", cause:(*errors.Err)(0xc820203c20), previous:(*errors.Err)(0xc82058a820), file:"github.com/juju/juju/worker/uniter/resolver/loop.go", line:66} ("could not acquire lock: cancelled acquiring mutex")
... error stack:
	github.com/juju/mutex/errors.go:12: cancelled acquiring mutex
	github.com/juju/juju/worker/uniter/uniter.go:543: 
	github.com/juju/juju/worker/uniter/operation/executor.go:74: could not acquire lock
	github.com/juju/juju/worker/uniter/resolver/loop.go:66: 
```

**Pull request merged**